src/Cli/dotnet/commands/VirtualProjectBuildingCommand.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System.Xml;
+using Microsoft.Build.Construction;
+using Microsoft.Build.Definition;
+using Microsoft.Build.Evaluation;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Logging;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Run;
+
+namespace Microsoft.DotNet.Tools;
+
+/// <summary>
+/// Used to build a virtual project file in memory to support <c>dotnet run file.cs</c>.
+/// </summary>
+internal sealed class VirtualProjectBuildingCommand
+{
+    public Dictionary<string, string> GlobalProperties { get; } = new(StringComparer.OrdinalIgnoreCase);
+    public required string EntryPointFileFullPath { get; init; }
+
+    public int Execute(string[] binaryLoggerArgs, LoggerVerbosity verbosity)
+    {
+        var binaryLogger = GetBinaryLogger(binaryLoggerArgs);
+        var consoleLogger = new ConsoleLogger(verbosity);
+        Dictionary<string, string?> savedEnvironmentVariables = new();
+        try
+        {
+            // Set environment variables.
+            foreach (var (key, value) in MSBuildForwardingAppWithoutLogging.GetMSBuildRequiredEnvironmentVariables())
+            {
+                savedEnvironmentVariables[key] = Environment.GetEnvironmentVariable(key);
+                Environment.SetEnvironmentVariable(key, value);
+            }
+
+            // Setup MSBuild.
+            ReadOnlySpan<ILogger> binaryLoggers = binaryLogger is null ? [] : [binaryLogger];
+            var projectCollection = new ProjectCollection(
+                GlobalProperties,
+                [.. binaryLoggers, consoleLogger],
+                ToolsetDefinitionLocations.Default);
+            var parameters = new BuildParameters(projectCollection)
+            {
+                Loggers = projectCollection.Loggers,
+                LogTaskInputs = binaryLogger is not null,
+            };
+            BuildManager.DefaultBuildManager.BeginBuild(parameters);
+
+            // Do a restore first (equivalent to MSBuild's "implicit restore", i.e., `/restore`).
+            // See https://github.com/dotnet/msbuild/blob/a1c2e7402ef0abe36bf493e395b04dd2cb1b3540/src/MSBuild/XMake.cs#L1838.
+            var restoreRequest = new BuildRequestData(
+                CreateProjectInstance(projectCollection, addGlobalProperties: static (globalProperties) =>
```

Comment:
cc @rainersigwald - potential use case for a public API to "do a restore"?

Reply:
That would be nice.

Reply:
Filed an issue for this one as well: https://github.com/dotnet/msbuild/issues/11519